### PR TITLE
[Attention][MLA] fp8_ds_mla for rope-free NoPE models (head_dim 512) on SM90 via zero-padded-rope envelope

### PR DESCRIPTION
**Pull request by Leoyzen**
## 🚀 Description

Enables `fp8_ds_mla` KV cache for rope-free NoPE sparse-MLA models (`qk_rope_head_dim=0`, head_dim 512, e.g. GLM-5.3-Flash) on **SM90/Hopper** via the FlashMLA sparse backend, served through a zero-padded 576/656B envelope.

Today on H200, GLM-5.3-Flash must either use bf16 KV (~2x memory) or the plain per-tensor-fp8 `FLASHINFER_MLA_SPARSE_SM90` backend, which returns no LSE and therefore cannot run DCP (decode context parallel) or compose with the MTP fused-draft path. The only SM90 sparse-MLA backend with `fp8_ds_mla` support (`FLASHMLA_SPARSE`, #47090) hard-rejects head_dim 512 because the FlashMLA kernels assume the fixed 576 geometry.

**Approach**: keep the 656B packed row layout unchanged and zero-pad the rope section:
- The rope bytes [640:768] of every cached token are written as bf16 zeros, and the query-side `q_pe` is zero-padded to 64 dims. Since RoPE is baked into the cache at write time, `q_pe · 0 = 0` exactly in IEEE arithmetic — the padding contributes nothing to any logit. This was verified end-to-end with a standalone kernel probe on H20 (SM90): output rel-err vs a bf16 NoPE-512 reference matches the fp8 block-quant noise floor (~2.5e-2), and a garbage-rope control diverges, proving the test has teeth.
- `FLASHMLA_SPARSE.get_supported_head_sizes()` → `[576, 512]`; `supports_combination` accepts 512 **only** for quantized DS-MLA dtypes (`fp8_ds_mla`/`nvfp4_ds_mla`), so bf16 NoPE-512 traffic keeps flowing to the FlashInfer/TRITON backends.
- A shared helper in `mla_attention.py` promotes the rope dims of rope-free MLA layers to the padded 576 geometry when the effective KV dtype is quantized DS-MLA on `FLASHMLA_SPARSE`, emitting persistent zero buffers (CUDA-graph capture stable). Models with rope dims > 0 are untouched.
- `SparseMLACommonMetadataBuilder` now derives its MLA dims from the instantiated runtime layers instead of the raw HF config, so the chunked-prefill workspace is sized from the promoted (576) geometry — without this, `cp_gather_and_upconvert_fp8_kv_cache` asserts `head_dim must be 576` (found in e2e).
- Backend selection: on SM90, `fp8_ds_mla` + head_size 512 now prefers `FLASHMLA_SPARSE` over `FLASHINFER_MLA_SPARSE_SM90`; plain `fp8` keeps FlashInfer as the documented stopgap (the two fp8 flavors remain distinguishable, no silent dtype reinterpretation).

**No csrc changes; no cache geometry changes.** The existing 576 traffic (DeepSeek family) is bit-identical. DCP mixed-batch LSE merge, MTP spec-decode reshape plumbing and CUDA-graph capture behavior are inherited unchanged (shapes and kernels are identical to the 576 path).

**Memory tradeoff (honest accounting)**: 656B/token vs a hypothetical "true NoPE fp8" 528B = +24% KV footprint; vs bf16 1152B/token it is still a 43% reduction. No SM90 kernel consumes a 512-only packed fp8 layout, so the padded envelope is the only way to get block-scaled fp8 KV on Hopper today.

### Not a duplicate of

- #47090 — FlashMLA fp8_ds_mla for SM90/SM100, but 576 geometry only; this PR extends it to NoPE-512 via the padded envelope.
- #53969 / #55277 — SM120 NoPE work on the FlashInfer SM120 backend (SM120 later moved to native `pe_dim=0`); this PR is the SM90 FlashMLA counterpart and cross-references them as siblings of the same technique.
- #54563 — fp8_ds_mla tile-scale reading; orthogonal.
- `FLASHINFER_MLA_SPARSE_SM90` plain-fp8 path — remains available and unchanged; it lacks LSE (no DCP) and per-tensor scaling.
- #54790 — fused draft metadata updates for sparse MLA backends (FlashMLA no-op + DSV3.2 indexer refresh + KPOOL_TAIL tail-cache mapping); orthogonal spec-decode change with no code dependency, but both were deployed and validated together on the same 8xH200 GLM-5.3-Flash stack (MTP K=3, full CUDA graphs).

## 📋 Tests

- `tests/v1/attention/test_flashmla_nope_sm90_backend_selection.py` — head-size/dtype acceptance matrix + SM90 priority (CPU-runnable): 14 passed, 4 skipped (no `vllm._C_stable_libtorch` on macOS dev host; CI-eligible on Linux CPU runner).
- `tests/v1/attention/test_flashmla_nope_sm90_fp8_ds_mla.py` — GPU suite (auto-skip without CUDA): zero-pad helper shape/zero/persistence, cache-write contract (rope bytes 640:768 bf16 zero, NoPE bytes match per-128-block fp8 reference), kernel equivalence vs bf16 NoPE-512 reference (rel-err < 0.08) with garbage-rope control, ragged mixed-batch (valid counts 0/1/partial, `-1` tails, all-invalid rows → (0, -inf)).
- Pre-commit (ruff-check/format, typos, mypy-3.10, SPDX) on all commits.

**E2E (8xH200, GLM-5.3-Flash, TP8/EP8, fp8_ds_mla, FLASHMLA_SPARSE, MTP K=3, chunked prefill, prefix caching, full CUDA graphs)**:
- Server starts cleanly; `Using FLASHMLA_SPARSE attention backend`; CUDA graph capture 51/51.
- MTP acceptance on natural English prompts: 78% (accepted 1717 / drafted 2202), in line with the pre-change baseline (85% on a different prompt mix), i.e. no acceptance regression from the envelope.

```
.venv/bin/python -m pytest tests/v1/attention/test_flashmla_nope_sm90_backend_selection.py -v
```

## 📝 Notes

- AI assistance was used for implementation and testing of this change; every line was reviewed and the e2e deployment was run and verified by the submitter.
- Follow-up (out of scope): a native 528B NoPE fp8 layout would remove the 128B padding but requires forking the vendored FlashMLA kernels; if SM120's native `pe_dim=0` (#55277) lands for SM90 later, this shim can be retired.

### CODE REVIEW

**Comment by hammad-hyde** [comment]
**Independent reproduction of the reference-axis bug in `test_nope_padded_envelope_matches_bf16_reference` (plus a second harness issue)**

We ported this PR onto a downstream tree (H100, SM90) and ran the envelope tests. Two findings, with reproduction:

**1. The bf16 reference normalizes over the wrong axis (the dominant failure).** `ref_probs = torch.softmax(ref_scores, dim=1)` normalizes `ref_scores` (shape `[num_tokens, num_heads, cache_positions]`) over **heads**, not cache tokens, so the reference is mathematically unrelated to attention. On our box the test then fails with `rel_err ≈ 1.815` (vs the ~2.5e-2 reported on H20) — the kernel output is actually correct: our end-to-end validation of this exact envelope path (GLM-5.3-Flash, `--kv-cache-dtype fp8` → zero-padded 656B envelope on `FLASHMLA_SPARSE`, TP=8 on 8×H100) passes GSM8K at **0.995 accuracy** (200q, 5-shot, vs the bf16 golden's 0.9900 — within run-to-run noise). Fix (already flagged by CodeRabbit): `torch.softmax(ref_scores, dim=-1)`.

**2. The hand-rolled `fm.flash_mla_with_kvcache` call diverges from the production path.** The test builds `tile_md/num_splits` and `indices` by hand rather than driving `FlashMLASparseImpl.forward_mqa`; on our tree the hand-built call produced output that diverges deterministically from the impl path even with the reference fixed. Recommendation: drive the kernel through `FlashMLASparseImpl.forward_mqa` with a minimal mock (fields: `num_heads`, `topk_indices_buffer`, `dcp_world_size=1`, `dcp_rank=0`, `need_to_return_lse_for_decode=False`, `index_group=None`, `fp8_decode_padded_heads`, `softmax_scale`, `_fp8_flash_mla_kernel`, `_convert_logical_to_physical_topk`) — the impl path is the contract the serving stack actually runs.

The ragged-batch test has a third, smaller harness gap already flagged in CodeRabbit's actionable list (the `kwargs["output"]` mock). Our environment: vllm dev e378275a8 (rebuilt from source), H100×8 (SM90), GLM-5.3-Flash 656B envelope, fp8_ds_mla. Happy to share the exact numbers/log snippets if useful.

*(AI-assisted reproduction; every claim above was verified by a human reviewer against logs and code.)*
